nemo/collections/asr/parts/numba/rnnt_loss/utils/cuda_utils/gpu_rnnt_kernel.py

```diff
@@ -397,3 +397,455 @@ def compute_grad_kernel(
+            # of the current step (t, u), normalized by the total log likelihood.
+            # Once the gradient has been calculated, scale it by `fastemit_lambda`, as in Equation 10.
+            if fastemit_lambda > 0.0 and u < U - 1:
+                fastemit_grad = fastemit_lambda * math.exp(
+                    alphas[col]  # alphas(t, u)
+                    + (denom[col] + acts[col * alphabet_size + labels[u]])
+                    - sigma
+                    + betas[col + 1]  # betas(t, u+1)
+                    + logpk  # log Pr(k|t, u)
+                    - logll[mb]  # total log likelihood for normalization
+                )
+            else:
+                fastemit_grad = 0.0
+
+            # Update the gradient of act[b, t, u, v] with the gradient from FastEmit regularization
+            grad = grad + fastemit_grad
+
+            # grad to last blank transition
+            # grad[b, T-1, U-1, v=blank] -= exp(alphas[b, t, u) + logpk - logll[b])
+            if (idx == blank_) and (t == T - 1) and (u == U - 1):
+                grad -= math.exp(alphas[col] + logpk - logll[mb])
+            else:
+                for i in range(num_big_blanks):
+                    if (idx == blank_ - 1 - i) and (t == T - big_blank_duration[i]) and (u == U - 1):
+                        grad -= math.exp(alphas[col] + logpk - logll[mb])
+
+            # grad of blank across t < T;
+            # grad[b, t<T-1, u, v=blank] -= exp(alphas[b, t, u] + logpk - logll[b] betas[b, t + 1, u])
+            if (idx == blank_) and (t < T - 1):
+                grad -= math.exp(alphas[col] + logpk - logll[mb] + betas[col + maxU])
+            else:
+                for i in range(num_big_blanks):
+                    if (idx == blank_ - 1 - i) and (t < T - big_blank_duration[i]):
+                        grad -= math.exp(alphas[col] + logpk - logll[mb] + betas[col + big_blank_duration[i] * maxU])
+
+            # grad of correct token across u < U;
+            # grad[b, t, u<U-1, v=label[u]] -= exp(alphas[b, t, u] + logpk - logll[b] + betas[b, t, u+1])
+            # Scale the gradient by (1.0 + FastEmit_lambda) in log space, then exponentiate
+            if (u < U - 1) and (idx == labels[u]):
+                # exp(log(1 + fastemit_lambda) + ...) is numerically more stable than
+                # multiplying (1.0 + fastemit_lambda) with result.
+                grad -= math.exp(math.log1p(fastemit_lambda) + alphas[col] + logpk - logll[mb] + betas[col + 1])
+
+            # update grads[b, t, u, v] = grad
+            grads[col * alphabet_size + idx] = grad
+
+            # clamp gradient (if needed)
+            if clamp > 0.0:
+                g = grads[col * alphabet_size + idx]
+                g = min(g, clamp)
+                g = max(g, -clamp)
+                grads[col * alphabet_size + idx] = g
+
+            # update internal index through the thread_buffer;
+            # until idx < V + 1, such that entire vocabulary has been updated.
```

Comment:
Just a note, incredible work on this. Nearly nothing to change other than docstrings

Reply:
Ha thanks.